### ISSUE TITLE
when the `IPv4` stack is used during build, invalid DNS names will cause a `free` of a `NULL` pointer in the SVM.

### ISSUE THREAD

**Issue by pmlopes** [oracle/graal#657]
The simple application:

```java
public class Main {
  public static void main(String[] args) {
    final Vertx vertx = Vertx.vertx();
    vertx.createHttpServer()
      .requestHandler(req -> req.response().end("Hello!"))
      .listen(Integer.getInteger("port", 8080));
  }
}
```

builds fine with `rc5` and with `rc6` fails with:

```
error: Detected a direct/mapped ByteBuffer in the image heap. A direct ByteBuffer has a pointer to unmanaged C memory, and C memory from the image generator is not available at image run time. A mapped ByteBuffer references a file descriptor, which is no longer open and mapped at run time. The object was probably created by a class initializer and is reachable from a static field. By default, all class initialization is done during native image building.You can manually delay class initialization to image run time by using the option --delay-class-initialization-to-runtime=<class-name>. Or you can write your own initialization methods and call them explicitly from your main entry point.
Detailed message:
Error: Detected a direct/mapped ByteBuffer in the image heap. A direct ByteBuffer has a pointer to unmanaged C memory, and C memory from the image generator is not available at image run time. A mapped ByteBuffer references a file descriptor, which is no longer open and mapped at run time. The object was probably created by a class initializer and is reachable from a static field. By default, all class initialization is done during native image building.You can manually delay class initialization to image run time by using the option --delay-class-initialization-to-runtime=<class-name>. Or you can write your own initialization methods and call them explicitly from your main entry point.
Trace:  object io.netty.buffer.UnpooledByteBufAllocator$InstrumentedUnpooledUnsafeNoCleanerDirectByteBuf
        object io.netty.buffer.UnreleasableByteBuf
        method io.netty.handler.codec.http.HttpObjectEncoder.encodeChunkedContent(ChannelHandlerContext, Object, long, List)
Call path from entry point to io.netty.handler.codec.http.HttpObjectEncoder.encodeChunkedContent(ChannelHandlerContext, Object, long, List): 
        at io.netty.handler.codec.http.HttpObjectEncoder.encodeChunkedContent(HttpObjectEncoder.java:196)
        at io.netty.handler.codec.http.HttpObjectEncoder.encode(HttpObjectEncoder.java:169)
        at io.vertx.core.http.impl.VertxHttpResponseEncoder.encode(VertxHttpResponseEncoder.java:32)
        at io.netty.handler.codec.MessageToMessageEncoder.write(MessageToMessageEncoder.java:88)
        at io.netty.channel.AbstractChannelHandlerContext.invokeWrite0(AbstractChannelHandlerContext.java:738)
        at io.netty.channel.AbstractChannelHandlerContext.invokeWrite(AbstractChannelHandlerContext.java:730)
        at io.netty.channel.AbstractChannelHandlerContext.access$1900(AbstractChannelHandlerContext.java:38)
        at io.netty.channel.AbstractChannelHandlerContext$AbstractWriteTask.write(AbstractChannelHandlerContext.java:1081)
        at io.netty.channel.AbstractChannelHandlerContext$AbstractWriteTask.run(AbstractChannelHandlerContext.java:1070)
        at io.netty.util.concurrent.GlobalEventExecutor$TaskRunner.run(GlobalEventExecutor.java:248)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.lang.Thread.run(Thread.java:748)
        at com.oracle.svm.core.posix.thread.PosixJavaThreads.pthreadStartRoutine(PosixJavaThreads.java:238)
        at com.oracle.svm.core.code.CEntryPointCallStubs.com_002eoracle_002esvm_002ecore_002eposix_002ethread_002ePosixJavaThreads_002epthreadStartRoutine_0028com_002eoracle_002esvm_002ecore_002eposix_002ethread_002ePosixJavaThreads_0024ThreadStartData_0029(generated:0)
```

While parsing netty's: `io.netty.handler.codec.http.HttpObjectEncoder`.

**Comment by pmlopes** [comment]
**Update**:

```
--delay-class-initialization-to-runtime=io.netty.handler.codec.http.HttpObjectEncoder,io.netty.handler.codec.http2.Http2CodecUtil,io.netty.handler.codec.http2.DefaultHttp2FrameWriter,io.netty.handler.codec.http.websocketx.WebSocket00FrameEncoder
```

Seems to allow things to build again... but then:

```
Server started at: http://localhost:8080
munmap_chunk(): invalid pointer
./build.sh: line 7:  5255 Aborted                 (core dumped) ./bitcoin-viewer
```

The application starts... the first line is debug logging and boom!

**Comment by Peter-B-Kessler** [comment]
Could you bundle your "simple application" in a repo with instructions as to how to compile, build an image, and run it?  Thanks.  Otherwise I have to guess at your "import" clauses, vertx versions, command lines, etc.  Or if you want to get your hands dirty: send me a `gdb` backtrace from the `core` file that was dumped.  (Please don't send me the `core` file itself.)  That might be diagnostic.

Almost certainly the problem is caused by a C pointer being passed to `free` that did not come from a `malloc` in the running image.  Direct ByteBuffers are one source of those pointers that we have tried to address with initialization at runtime, but we might have missed other sources of such pointers.  Having to build a list of classes to initialize at runtime, by finding them (or not!) one at a time at runtime, seems like a brittle way to build a robust application.

**Comment by pmlopes** [comment]
@Peter-B-Kessler I've started looking at the issue again after a night of sleep and finally isolated it to the core JDK/SVM, it is not related to Netty or Vert.x :smile: 

This is the basic reproducer (no dependencies):

```java
public class Main {
  public static void main(String[] args) throws Exception {
    java.net.InetAddress.getByName("oopsydaisyidontexist");
  }
}
```

Now I was building this image with:

```
native-image \
  --no-server \
  -Djava.net.preferIPv4Stack=true \
  -H:+ReportUnsupportedElementsAtRuntime \
  -jar issue.jar
```

This would render the `segfault`:

```
  RSP 00007f0946ffd480 RIP 0000000000a05885  [image code] com.oracle.svm.core.posix.headers.Netdb.freeaddrinfo(Netdb.java)
  RSP 00007f0946ffd480 RIP 0000000000a05885  [image code] com.oracle.svm.core.posix.Target_java_net_Inet4AddressImpl.lookupAllHostAddr(PosixJavaNetSubstitutions.java:2886)
  RSP 00007f0946ffd5b0 RIP 0000000000a0a00f  [image code] java.net.InetAddress$2.lookupAllHostAddr(InetAddress.java:928)
  RSP 00007f0946ffd5b0 RIP 0000000000a0a00f  [image code] java.net.InetAddress.getAddressesFromNameService(InetAddress.java:1323)
  RSP 00007f0946ffd680 RIP 0000000000a0c5a5  [image code] java.net.InetAddress.getAllByName0(InetAddress.java:1276)
  RSP 00007f0946ffd6c0 RIP 0000000000a0b2bb  [image code] java.net.InetAddress.getAllByName(InetAddress.java:1192)
  RSP 00007f0946ffd730 RIP 0000000000a0c8bb  [image code] java.net.InetAddress.getAllByName(InetAddress.java:1126)
  RSP 00007f0946ffd730 RIP 0000000000a0c8bb  [image code] java.net.InetAddress.getByName(InetAddress.java:1076)
```
Now, the interesting part is that if I remove the property:

```
  -Djava.net.preferIPv4Stack=true
```

from the native image, then I get a plain Java Exception:

```
Exception in thread "main" java.net.UnknownHostException: oopsydaisyidontexist
        at java.lang.Throwable.<init>(Throwable.java:265)
        at java.lang.Exception.<init>(Exception.java:66)
        at java.io.IOException.<init>(IOException.java:58)
        at java.net.UnknownHostException.<init>(UnknownHostException.java:47)
        at com.oracle.svm.core.posix.Target_java_net_Inet6AddressImpl.lookupAllHostAddr(PosixJavaNetSubstitutions.java:3257)
        at java.net.InetAddress$2.lookupAllHostAddr(InetAddress.java:928)
        at java.net.InetAddress.getAddressesFromNameService(InetAddress.java:1323)
        at java.net.InetAddress.getAllByName0(InetAddress.java:1276)
        at java.net.InetAddress.getAllByName(InetAddress.java:1192)
        at java.net.InetAddress.getAllByName(InetAddress.java:1126)
        at java.net.InetAddress.getByName(InetAddress.java:1076)
        at g.Main.main(Main.java:6)
        at com.oracle.svm.core.JavaMainWrapper.run(JavaMainWrapper.java:163)
```

Which can be handled by my application.

To sum up, it seems that when the `IPv4` stack is used during build, invalid DNS names will cause a `free` of a `NULL` pointer in the SVM.

**Comment by Peter-B-Kessler** [comment]
I hope that https://github.com/oracle/graal/commit/7c850c39d4580c85d2eed9a9b13b80398554689f fixes the problem.  You will have to [build from source](https://github.com/oracle/graal/tree/master/substratevm#quick-start) until our next candidate release.

I wrote tests for `InetAddress.getByName(String)`: one for a hostname that exists, and one for a hostname that does not exist.  And to try out our process for writing tests.  You can run the tests with
```
graal/substratevm $ mx native-unittest TestInetAddress
Building junit image for matching: com.oracle.svm.test.TestInetAddress
[svmjunit:98518]    classlist:   2,501.00 ms
[svmjunit:98518]        (cap):   1,642.27 ms
[svmjunit:98518]        setup:   3,939.65 ms
[svmjunit:98518]   (typeflow):   8,333.80 ms
[svmjunit:98518]    (objects):   2,591.58 ms
[svmjunit:98518]   (features):      80.48 ms
[svmjunit:98518]     analysis:  11,173.00 ms
[svmjunit:98518]     universe:     410.24 ms
[svmjunit:98518]      (parse):   3,703.78 ms
[svmjunit:98518]     (inline):   2,679.54 ms
[svmjunit:98518]    (compile):  17,018.14 ms
[svmjunit:98518]      compile:  26,647.25 ms
[svmjunit:98518]        image:   1,243.73 ms
[svmjunit:98518]        write:     798.12 ms
[svmjunit:98518]      [total]:  46,792.95 ms
SVMJUnitCore
JUnit version 4.12
com.oracle.svm.test.TestInetAddress started (1 of 1)
  testGetByNameExistingHost: Passed
  testGetByNameNonExistingHost: Passed
com.oracle.svm.test.TestInetAddress finished
Time: 0.005

OK (2 tests)
```
